src/dataset.jl

```diff
@@ -172,13 +172,114 @@ end
 )::Nothing
-    destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    if nraster(dataset) > 0 && nlayer(dataset) > 0
+        error("Writing datasets with raster and vector data is not supported.")
+    elseif nraster(dataset) > 0
+        destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    elseif nlayer(dataset) > 0
+        writelayers(dataset, filename; kwargs...)
+    end
+    return nothing
+end
+
+"""
+    writelayers(ds, filename; driver, options=[""], layer_options=[""], chunksize=20000)
+
+Writes the vector dataset to the designated filename. The options are passed to the newly created dataset and
+have to be given as a list of strings in KEY=VALUE format. The chunksize controls the number of features written 
+in each database transaction, e.g. for SQLite. This function can also be used to copy datasets on disk.
+
+Currently working drivers: FlatGeobuf, GeoJSON, GeoJSONSeq, GML, JML, KML, MapML, ESRI Shapefile, SQLite
+GPKG should mostly work, too
+
+### Parameters
+*`dataset`  The source dataset
+*`filename` The file name to write to
+
+### Keyword arguments
+`driver`           The driver to use, you have to manually select the right driver for the file extension you wish
+`options`   A vector of strings containing KEY=VALUE pairs for driver-specific creation options
+`layer_options`    Driver specific options for layer creation
+`chunksize`        Number of features to write in one database transaction. Neglected when `use_gdal_copy` is true.
+`use_gdal_copy`    Set this to true if you encounter errors when writing. Set to false for higher speed.
+
+### Returns
+nothing
+"""
+function writelayers(dataset::AbstractDataset,
+                     filename;
+                     driver=getdriver(dataset),
+                     options::Vector{String}=[""], 
+                     layer_options::Vector{String}=[""],
+                     chunksize=20000,
+                     use_gdal_copy=false)
+    create(filename; driver=driver, options=options) do target
+        for layeridx in 0:nlayer(dataset)-1
+            sourcelayer = getlayer(dataset, layeridx)
+            sourcelayerdef = layerdefn(sourcelayer)
+
+            if use_gdal_copy
+                copy(sourcelayer; dataset=target, name=getname(sourcelayer))
+            else
+                createlayer(name=getname(sourcelayer),
+                            dataset=target,
+                            geom=getgeomtype(sourcelayer),
+                            spatialref=getspatialref(sourcelayer),
+                            options=layer_options) do targetlayer
+
+                    targetlayerdef = layerdefn(targetlayer)
+                    
+                    # work around GPKG driver quirks
+                    # GPKG still has issues with empty geometry field names,
+                    # especially when the field name is empty
+                    if shortname(driver) == "GPKG"
+                        deletegeomdefn!(targetlayerdef, 0)
```

Comment:
So I think this is technically [not allowed by GDAL](https://github.com/OSGeo/gdal/blob/023d25cf4508e6a4fdae8deaf399bc8fc00825cd/ogr/ogrfeaturedefn.cpp#L839): https://github.com/yeesian/ArchGDAL.jl/blob/04d067e15b2d9fea2987cb2dd2e130f866dda73c/src/ogr/featurelayer.jl#L550-L558.

I'm curious it's working for GPKG but I think it might be worth filing an issue with https://github.com/OSGeo/gdal to ask about the situation with GPKG and see what should be done.

Reply:
I think it's ok to use it in this case, for following practical reasons: 1) The layer is empty by the time this is called, as requested by the docstring. 2) The alternative mentioned in the docstring is not implemented yet. 3) It is used in other places of the gdal codebase, e.g. in the HANA driver. 4) It makes things work for one driver more, where it otherwise wouldn't.

Reply:
I just found a better workaround by setting the geometry column name to the already source geom col name.